whole-tale/girder_wholetale
# Add Tale build endpoint

Fixes #206, implementing endpoints for per-Tale environment customization via `repo2docker` in conjunction with https://github.com/whole-tale/gwvolman/pull/43.

This PR does the following:
* Adds the `/tale/{id}/build` endpoint which calls the new `build_tale_image` gwvolman task (https://github.com/whole-tale/gwvolman/pull/43)
* Adds `updateBuildStatus` event handler to add `tale['imageInfo']` containing image digest, build jobId and status, including event triggering
* Adds `updateWorkspaceModTime` event handler to detect save/create/delete events on files in the workspace
* Removes `/image/{id}/build` endpoint along with the `Recipe` object and related methods/tests
* Adds unit/integration tests

To test
* `https://github.com/whole-tale/deploy-dev -b repo2docker`  
  * This changes the Image initialization process in `setup_girder.py` to create new image objects
* Checkout the following branches:
  *  `src/wholetale` `206-tale-build-endpoint`
  * `src/gwvolman` `40-tale-build-repo2docker`.
* Login to Dashboard.  Confirm the compose page has 4 environments: Jupyter (Classic), JupyterLab, RStudio, RStudio (rocker/geospatial)
* Compose a new tale called "Jupyter Tale" using the Jupyter (Classic) environment
  * Confirm that the tale instance starts with the classic Jupyter notebook
  * Inspect `celery_worker` logs or logs for `Build Tale Image` job.  Confirm that image was built)
  * In your workspace, create a new file called `apt.txt` with package `byacc`
  * Using the dashboard or Swagger API page, call `/tale/{id}/build` for the Tale
  * Inspect `celery_worker` logs or logs for `Build Tale Image` job to confirm that the image was built and `byacc` package installed
  * Stop and restart the instance (or use new restart endpoint)
  * Using the Jupyter terminal, confirm `byacc` is installed by running `byacc`
  * Modify `apt.txt` adding `vim` and repeat steps above to confirm that `vim` is installed and can be used in new image.
  * Delete `apt.txt` and repeat the steps above to confirm that neither `byacc` nor `vim` are installed in the new image
 * Make no changes and rebuild the image.  Inspect the celery logs or job logs to confirm that the image was not built since the workspace was unchanged.
* Compose a new tale called "JupyterLab Tale". Confirm that the interface displays JupyterLab.
* Compose a new tale called "Rocker Tale" using the RStudio (rocker/geospatial) environment
  * To confirm this is the rocker-based image via, open Tools > Shell and run `gdal-config`
  * Create a new file called install.R with the line `install.packages("mvtnorm")`
  * Using the Swagger API page, call `/tale/{id}/build` for the associated Tale
  * Use `docker logs -f celery_worker` to confirm that the image builds without error
  * Stop and restart the instance.
  * Using the R console, confirm the package is installed via `library(mvtnorm)`
* Compose a new tale called "RStudio Tale" using the RStudio environment
  * Repeat the steps for the Rocker tale, note that gdal-config is not present
  * Note: the RStudio build currently takes a very long time
